File: 알고리즘 정리/힙/문제풀이/디스크 컨트롤러.py
##################################################################
import heapq
from math import floor
def solution(jobs):
    jobs.sort() #시작 시간 기준 오름차순 정렬
    answer = 0
    start = 0 #작업 진행시작시간
    minHeap=[]
    for job in jobs:
        # 다음 작업중 가장 빠른 시작 시간보다 start가 작으면 이전 작업들중 작업 시간이 가장 적게 걸리는 것 우선으로 작업 진행
        while minHeap and job[0]>start:
            temp = heapq.heappop(minHeap)
            answer += start - temp[1] + temp[0]
            start += temp[0]
        # 다음 작업 시작 시간이 start보다 큰 경우는 start를 해당 작업 시작 시간으로 설정
        if not minHeap and job[0]>start:
            start = job[0]
        heapq.heappush(minHeap,[job[1],job[0]])

    #남은 작업중 시간이 가장 적게걸리는 순으로 작업 진행
    while minHeap:
        temp = heapq.heappop(minHeap)
        answer += start - temp[1] + temp[0]
        start += temp[0]
    return floor(answer/len(jobs))
##################################################################

jobs=[[0, 3], [1, 9], [2, 6]]
print(solution(jobs))

File: 알고리즘 정리/힙/문제풀이/디스크 컨트롤러.cpp
#include<iostream>

////////////////////////////////////////////////////////
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

struct cmp
{
	bool operator()(vector<int> a, vector<int> b)
	{
		return a.at(1) > b.at(1);
	}
};

int solution(vector<vector<int>> jobs) {
	int answer = 0, i = 0, time = 0;
	sort(jobs.begin(), jobs.end());
	priority_queue<vector<int>, vector<vector<int>>, cmp> pq;
	while (i < jobs.size() || !pq.empty()) 
	{
		if (jobs.size() > i && time >= jobs[i][0])
		{
			pq.push(jobs[i]);
			i++;
			continue;
		}
		if (!pq.empty())
		{
			time += pq.top()[1];
			answer += time - pq.top()[0];
			pq.pop();
		}
		else 
			time = jobs[i][0];
	}
	return answer / jobs.size();
}
////////////////////////////////////////////////////////

int main()
{
	cin.sync_with_stdio(0);
	cin.tie(0); cout.tie(0);
	vector<vector<int>> jobs = { {0,3},{1,9},{2,6} };
	cout << solution(jobs);
	return 0;
}